jdbc/driver/src/main/java/io/tidb/bigdata/jdbc/LoadBalancingDriver.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright 2020 TiDB Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.tidb.bigdata.jdbc;
+
+import static io.tidb.bigdata.jdbc.ExceptionHelper.call;
+import static io.tidb.bigdata.jdbc.ExceptionHelper.stringify;
+import static io.tidb.bigdata.jdbc.ExceptionHelper.uncheckedCall;
+import static java.util.Objects.requireNonNull;
+
+import io.tidb.bigdata.jdbc.impl.DiscovererImpl;
+import io.tidb.bigdata.jdbc.impl.RandomShuffleUrlMapper;
+import io.tidb.bigdata.jdbc.impl.RoundRobinUrlMapper;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.sql.Connection;
+import java.sql.Driver;
+import java.sql.DriverPropertyInfo;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.logging.Logger;
+
+public class LoadBalancingDriver implements Driver {
+  private static final Logger logger = Logger.getLogger(LoadBalancingDriver.class.getName());
+
+  private static final String MYSQL_URL_PREFIX = "jdbc:mysql://";
+  /**
+   * The value is set by {@link System#setProperty(String, String)}
+   */
+  private static final String TIDB_URL_MAPPER = "tidb.jdbc.url-mapper";
+  /**
+   * The value is set by {@link System#setProperty(String, String)}
+   */
+  private static final String URL_PROVIDER = "url.provider";
+  /**
+   * The value is set by {@link System#setProperty(String, String)}
+   */
+  private static final String TIDB_MIN_DISCOVER_INTERVAL_KEY = "tidb.jdbc.min-discovery-interval";
+  private static final long TIDB_MIN_DISCOVER_INTERVAL = 1000;
+  /**
+   * The value is set by {@link System#setProperty(String, String)}
+   */
+  private static final String TIDB_MAX_DISCOVER_INTERVAL_KEY = "tidb.jdbc.min-discovery-interval";
```

Comment:
shouldn't the value be `tidb.jdbc.max-discovery-interval`?

Reply:
good catch!